### PR TITLE
Java: Add support for synthetic callables with flow summaries and model Stream.collect

### PR DESCRIPTION
**Pull request by aschackmull**
This covers basic support `Stream.collect`. There are more patterns that could be covered (see e.g. the qltest - currently the collector is required to be directly in the argument position), but adding support for more collectors would require some additional plumbing to connect the collector arguments to the `collect` call.

### CODE REVIEW

File: java/ql/lib/semmle/code/java/frameworks/Stream.qll [hunk 1]
```diff
@@ -1,6 +1,101 @@
...
+
+  predicate hasName(string name) { this.getMethod().hasName(name) }
+}
+
+private class CollectToContainer extends SyntheticCallable {
+  CollectToContainer() { this = "java.util.stream.collect()+Collectors.[toList,...]" }
+
+  override Call getACall() {
+    result
+        .(CollectCall)
```
Comment (hvitved) [inline java/ql/lib/semmle/code/java/frameworks/Stream.qll:27]: Could change return type to `CollectCall` instead.

File: java/ql/lib/semmle/code/java/frameworks/Stream.qll [hunk 2]
```diff
@@ -1,6 +1,101 @@
...
+    input = SummaryComponentStack::elementOf(SummaryComponentStack::qualifier()) and
+    output = SummaryComponentStack::elementOf(SummaryComponentStack::return()) and
+    preservesValue = true
+  }
+}
+
+private class CollectToJoining extends SyntheticCallable {
+  CollectToJoining() { this = "java.util.stream.collect()+Collectors.joining" }
+
+  override Call getACall() { result.(CollectCall).getArgument(0).(Collector).hasName("joining") }
```
Comment (hvitved) [inline java/ql/lib/semmle/code/java/frameworks/Stream.qll:48]: Same about return type.

File: java/ql/lib/semmle/code/java/frameworks/Stream.qll [hunk 3]
```diff
@@ -1,6 +1,101 @@
...
+}
+
+private class CollectToGroupingBy extends SyntheticCallable {
+  CollectToGroupingBy() {
+    this = "java.util.stream.collect()+Collectors.[groupingBy(Function),...]"
+  }
+
+  override Call getACall() {
+    exists(Method m |
+      m = result.(CollectCall).getArgument(0).(Collector).getMethod() and
```
Comment (hvitved) [inline java/ql/lib/semmle/code/java/frameworks/Stream.qll:68]: same

**Review by hvitved — COMMENTED** [review]
Approach LGTM. Some minor comments.

File: java/ql/lib/semmle/code/java/frameworks/Stream.qll (hunk 1 above)
Comment (aschackmull) [reply java/ql/lib/semmle/code/java/frameworks/Stream.qll:27]: I'm not sure I like that particular asymmetric QL language feature, so I think I'll keep it as-is.

**Comment by aschackmull** [comment]
dca looks good.

**Review by hvitved — APPROVED** [review]

**Pull request merged**